2016-11-17-mvcc-in-tikv.md

```diff
@@ -70,55 +70,57 @@ impl Storage {
 
 ### MVCC
 
-Here comes the core of the transaction model of TiKV, which called 2-Phase Commit powered by MVCC. There are two stages in one transaction.
+#### Column family
 
-#### Prewrite
+Compared with Percolator where the information such as Lock is stored by adding extra column to a specific row, TiKV uses a column family (CF) in RocksDB to handle all the information related to Lock. To be specific, TiKV stores the Key-Values, Locks and Writes information in `CF_DEFAULT`, `CF_LOCK`, `CF_WRITE`.
 
-1. Select one row as the primary row, the others as the secondary rows.
-2. [Lock the primary row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L80). Before locking, it will check [whether there is other locks on this row](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L71) or whether there are some commits located after startTS. These two situations will lead to conflicts.If any of themhappens, [rollback](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L115) will be called.
-3. Repeat the  operations on secondary row.
+All the values of the CF are encoded as following:
 
-#### Commit
+|  | default | lock | write |
+| --- | --- | --- | --- |
+| **key** | z{encoded_key}{start_ts(desc)} | z{encoded_key} | z{encoded_key}{commit_ts(desc)} |
+| **value** | {value} | {flag}{primary_key}{start_ts(varint)} | {flag}{start_ts(varint)} |
 
-1. Write to the `CF_WRITE` with commitTS.
-2. Delete the corresponding lock.
+More details can be found [here](https://github.com/pingcap/tikv/issues/1077).
 
-#### Rollback
+#### Transaction model
 
-[Rollback](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L115) is called when there are conflicts during the `prewrite`.
+Here comes the core of the transaction model for TiKV, which is MVCC powered by 2-phase commit. There are two stages in one transaction:
 
-#### Garbage collector
-It is easy to predict that there will be more and more MVCC versions if there is no [Garbage Collector](https://github.com/pingcap/tikv/blob/master/src/storage/mvcc/txn.rs#L143) to remove the invalid versions. But we cannot just simply removeall the versions before a safe point. Since there maybe only one version for a key, it will be kept. In `TiKV`, if there is  any `Put` or `Delete` before the safe point, then all the latter writes can be deleted, otherwise only `Delete`, `Rollback` and `Lock` will be deleted.
+- **Prewrite**
 
-# TiKV-Ctl for MVCC
```

Comment:
Removed the topic of **TiKV-Ctl for MVCC** since this feature has evolved a lot since this article.